### ISSUE TITLE
unsafe.JObject equality

### ISSUE THREAD

**Issue by dwijnand** [mdedetrich/scala-json-ast#10]
If this is intended it should be declared loudly, because personally I find it very unintuitive in Scala:

```scala
scala> import scala.json.ast._
import scala.json.ast._

scala> unsafe.JObject() == unsafe.JObject()
res0: Boolean = false
```

**Comment by mdedetrich** [comment]
Let me have a look, its probably due to reference equality since the underlying datastructure is an array (this is actually something that should be discussed in general from a design POV)

**Comment by mdedetrich** [comment]
I will live this open for discussion, however documentation has been updated to reflect the fact that the equality is by reference. Since one of the main use cases for unsafe is performance, there are arguments for leaving it the way it is

**Comment by farmdawgnation** [comment]
I'm not too wild about "unsafe" also being unsafe to use `==` on. Is there some way for `==` to just be a more expensive, but still accurate, operation in the unsafe AST?

**Comment by mdedetrich** [comment]
> I'm not too wild about "unsafe" also being unsafe to use == on. Is there some way for == to just be a more expensive, but still accurate, operation in the unsafe AST?

There is but it then means you can't do reference equality anymore (afaik), and one of the design goals of unsafe is to be fast by default.

Obviously you can do structural equality with the Array, there is java.util.Arrays.equals which will do a proper comparison that 2 arrays are equal. It does get a bit more complicated with `JObject` though since you have to do this logic recursively. I am thinking that it may make sense to provide a helper method to do this, I had to do it in tests to verify correctness anyways

**Comment by dwijnand** [comment]
`eq` is available for reference equality

**Comment by mdedetrich** [comment]
Okay, so there is an argument for making it do structural equality by default

**Comment by mdedetrich** [comment]
Fixed in 1.0.0-M5
